### PR TITLE
Move SigSet, TimeVal to libc, add TimeSpec and pselect syscall.

### PR DESCRIPTION
**Pull request by abbradar**
Part and extension of #270 and should be merged after #272. Maybe we want to split this across multiple PRs too (although this time these things are dependent on each other).

### CODE REVIEW

**Comment by abbradar** [comment]
I guess we should use proper `time_t` instead of `u64` and stuff -- I'll fix this later.

**Comment by kamalmarhubi** [comment]
When you do, could you also rebase on top of master now that #272 was merged?

**Comment by kamalmarhubi** [comment]
(Or drop the duplicate commits)

**Comment by abbradar** [comment]
Just to give a signal -- I remember about those and I'll continue working at them when I get more time. Sorry for the delay!

**Comment by kamalmarhubi** [comment]
@abbradar no worries! And thanks again for your contributions and interest! :-)

**Comment by abbradar** [comment]
Argh, I've just now rebased it atop master and now I cannot find old comments which you've done.... @kamalmarhubi , can you please review this once more time! Or, even better, is there any way to actually get those old comments when you've force-pushed?

**Comment by fiveop** [comment]
I checked my emails. There were no other comments of @kamalmarhubi on this PR.

**Comment by abbradar** [comment]
I was talking about his old comments. Said that, I was able to find them in the mailbox archive -- sorry for the noise! I'll go through them shortly.

**Comment by abbradar** [comment]
I've moved `TimeVal` and `TimeSpec` to the newtype pattern. I think those are the only relevant comments now; if I'm mistaken, please correct me!

File: src/sys/time.rs [hunk 1]
```diff
@@ -208,17 +216,16 @@ mod test {
     #[test]
     pub fn test_time_val() {
         assert!(TimeVal::seconds(1) != TimeVal::zero());
-        assert_eq!(TimeVal::seconds(1) + TimeVal::seconds(2), TimeVal::seconds(3));
-        assert_eq!(TimeVal::minutes(3) + TimeVal::seconds(2),
-                   TimeVal::seconds(182));
+        assert!(TimeVal::seconds(1) + TimeVal::seconds(2) == TimeVal::seconds(3));
+        assert!(TimeVal::minutes(3) + TimeVal::seconds(2) == TimeVal::seconds(182));
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: `assert_eq` is preferable because it prints the left and right hand sides

File: src/sys/time.rs [hunk 2]
```diff
@@ -208,17 +216,16 @@ mod test {
...
     }
 
     #[test]
     pub fn test_time_val_neg() {
         let a = TimeVal::seconds(1) + TimeVal::microseconds(123);
         let b = TimeVal::seconds(-1) + TimeVal::microseconds(-123);
 
-        assert_eq!(a, -b);
+        assert!(a == -b);
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: same re `assert_eq`

File: src/sys/time.rs [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
+        &self.0
+    }
+}
 
-const MIN_SECONDS: i64 = -MAX_SECONDS;
+impl AsMut<timeval> for TimeVal {
+    fn as_mut(&mut self) -> &mut timeval {
+        &mut self.0
+    }
+}
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: oh I like having `AsRef` and `AsMut` impls for newtypes. making a mental note to go back and look at what exists already...

File: src/sys/time.rs [hunk 4]
```diff
@@ -147,32 +147,40 @@ impl ops::Div<i32> for TimeVal {
...
 }
 
+impl PartialEq for TimeVal {
+    fn eq(&self, rhs: &TimeVal) -> bool {
+        self.0.tv_sec == rhs.0.tv_sec && self.0.tv_usec == rhs.0.tv_usec
+    }
+}
+
+impl Eq for TimeVal { }
+
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: could you also implement `PartialOrd`, `Ord`, and `Debug`, which we lost in move to libc. I figure that `Debug` could print like `TimeVal(timeval { tv_sec: 10, tv_usec: 20 })` as the derived impl would have. I copied the definitions to playground to demonstrate what `derive(Debug)` would do: https://is.gd/jK1j9K

File: src/sys/time.rs [hunk 5]
```diff
@@ -1,24 +1,24 @@
...
-    pub tv_usec: suseconds_t,
-}
+use libc::{time_t, suseconds_t, timeval};
 
 const MICROS_PER_SEC: i64 = 1_000_000;
 const SECS_PER_MINUTE: i64 = 60;
 const SECS_PER_HOUR: i64 = 3600;
 
-#[cfg(target_pointer_width = "64")]
-const MAX_SECONDS: i64 = (::std::i64::MAX / MICROS_PER_SEC) - 1;
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: why was this removed?

File: src/sys/time.rs [hunk 6]
```diff
@@ -181,6 +182,191 @@ impl PartialEq for TimeVal {
...
 impl Eq for TimeVal { }
 
+#[derive(Clone, Copy)]
+pub struct TimeSpec(pub timespec);
+
+impl AsRef<timespec> for TimeSpec {
+    fn as_ref(&self) -> &timespec {
+        &self.0
+    }
+}
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: `AsMut`?

File: src/sys/time.rs [hunk 7]
```diff
@@ -181,6 +182,191 @@ impl PartialEq for TimeVal {
...
+
+impl TimeSpec {
+    #[inline]
+    pub fn zero() -> TimeSpec {
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: I'd prefer to hold off adding all these methods until #277 is decided, which would likely mostly be interconversion with [`std::time::Duration`](http://doc.rust-lang.org/std/time/struct.Duration.html).

File: src/sys/time.rs [hunk 8]
```diff
@@ -181,6 +182,191 @@ impl PartialEq for TimeVal {
...
+    }
+}
+
+impl PartialEq for TimeSpec {
+    fn eq(&self, rhs: &TimeSpec) -> bool {
+        self.0.tv_sec == rhs.0.tv_sec && self.0.tv_nsec == rhs.0.tv_nsec
+    }
+}
+
+impl Eq for TimeSpec { }
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: `Ord`, `PartialOrd`, `Debug`?

File: src/sys/time.rs [hunk 9]
```diff
@@ -236,4 +422,29 @@ mod test {
         assert_eq!(TimeVal::microseconds(42).to_string(), "0.000042 seconds");
         assert_eq!(TimeVal::seconds(-86401).to_string(), "-86401 seconds");
     }
+
+    #[test]
+    pub fn test_time_spec() {
+        assert!(TimeSpec::seconds(1) != TimeSpec::zero());
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: prefer `assert_eq`

File: src/sys/time.rs [hunk 10]
```diff
@@ -236,4 +422,29 @@ mod test {
...
+        assert!(TimeSpec::seconds(1) + TimeSpec::seconds(2) == TimeSpec::seconds(3));
+        assert!(TimeSpec::minutes(3) + TimeSpec::seconds(2) == TimeSpec::seconds(182));
+    }
+
+    #[test]
+    pub fn test_time_spec_neg() {
+        let a = TimeSpec::seconds(1) + TimeSpec::nanoseconds(123);
+        let b = TimeSpec::seconds(-1) + TimeSpec::nanoseconds(-123);
+
+        assert!(a == -b);
```
Comment (kamalmarhubi) [inline src/sys/time.rs]: prefer `assert_eq`

File: src/sys/select.rs [hunk 11]
```diff
@@ -1,85 +1,75 @@
...
-
-#[cfg(not(any(target_os = "macos", target_os = "ios")))]
-#[repr(C)]
-#[derive(Clone)]
-pub struct FdSet {
-    bits: [u64; FD_SETSIZE as usize / 64]
+impl AsRef<fd_set> for FdSet {
+    fn as_ref(&self) -> &fd_set {
+        &self.set
+    }
```
Comment (kamalmarhubi) [inline src/sys/select.rs]: `AsMut`?

File: src/sys/select.rs [hunk 12]
```diff
@@ -1,85 +1,75 @@
...
 }
 
-#[cfg(not(any(target_os = "macos", target_os = "ios")))]
-const BITS: usize = 64;
+pub const FD_SETSIZE: RawFd = libc::FD_SETSIZE as RawFd;
```
Comment (kamalmarhubi) [inline src/sys/select.rs]: this should be a `usize` as in libc.

File: src/sys/select.rs [hunk 13]
```diff
@@ -1,85 +1,75 @@
...
+        unsafe {
+            libc::FD_CLR(fd, &mut self.set as *mut _);
+        }
     }
 
-    pub fn contains(&mut self, fd: RawFd) -> bool {
-        let fd = fd as usize;
-        self.bits[fd / BITS] & (1 << (fd % BITS)) > 0
+    pub fn contains(&self, fd: RawFd) -> bool {
+        assert!(fd >= 0 && fd < FD_SETSIZE, "RawFd out of bounds");
```
Comment (kamalmarhubi) [inline src/sys/select.rs]: we can probably just return false in this case, while preserving expected meaning?

File: src/sys/select.rs [hunk 14]
```diff
@@ -1,85 +1,75 @@
...
+        unsafe {
+            // We require `transmute` here because FD_ISSET wants a mutable pointer,
+            // when in fact it doesn't mutate.
```
Comment (kamalmarhubi) [inline src/sys/select.rs]: :+1: for the comment, thanks :-)

**Comment by kamalmarhubi** [comment]
I made some comments. I do think though that it would be easier to review further changes if each commit was its own PR.

**Comment by kamalmarhubi** [comment]
(And thanks for patience, and for coming back to this PR!)

File: src/sys/time.rs (hunk 4 above)
Comment (kamalmarhubi) [inline src/sys/time.rs]: TIL: [`std::fmt::Formatter::debug_struct`](http://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct)

**Comment by Susurrus** [comment]
@abbradar I think most of these changes have already been done or are not decided on, but the addition of `pselect()` would be most welcome. Are you still interested in merging that? If you can remove the other 3 commits here, we'd be all set!

**Comment by abbradar** [comment]
I can rebase of course but I don't have much left over from my old project from that time so I cannot test it responsibly now.

**Comment by Susurrus** [comment]
@abbradar That's fair and I'm going to close this then. Thanks anyways for submitting this PR, sorry it languished so long.
